src/main/telemetry/srxl.c

```diff
@@ -172,31 +189,144 @@ void srxlFrameFlightPackCurrent(sbuf_t *dst)
+#define SPEKTRUM_SRXL_DEVICE_TEXTGEN_COLS (13)  // Text Generator COLS
+
+/*
+typedef struct
+{
+    UINT8		identifier;
+    UINT8		sID;		       // Secondary ID
+    UINT8		lineNumber;	       // Line number to display (0 = title, 1-8 for general, 254 = Refresh backlight, 255 = Erase all text on screen)
+    char		text[13];	       // 0-terminated text when < 13 chars
+} STRU_SPEKTRUM_SRXL_TEXTGEN;
+*/
+
+#if ( SPEKTRUM_SRXL_TEXTGEN_BUFFER_COLS > SPEKTRUM_SRXL_DEVICE_TEXTGEN_COLS )
+static char srxlTextBuff[SPEKTRUM_SRXL_TEXTGEN_BUFFER_ROWS][SPEKTRUM_SRXL_TEXTGEN_BUFFER_COLS];
+static bool lineSent[SPEKTRUM_SRXL_TEXTGEN_BUFFER_ROWS];
+#else
+static char srxlTextBuff[SPEKTRUM_SRXL_DEVICE_TEXTGEN_ROWS][SPEKTRUM_SRXL_DEVICE_TEXTGEN_COLS];
+static bool lineSent[SPEKTRUM_SRXL_DEVICE_TEXTGEN_ROWS];
+#endif
+
+//**************************************************************************
+// API Running in external client task context. E.g. in the CMS task 
+int spektrumTmTextGenPutChar(uint8_t col, uint8_t row, char c)
+{
+    if (row < SPEKTRUM_SRXL_TEXTGEN_BUFFER_ROWS && col < SPEKTRUM_SRXL_TEXTGEN_BUFFER_COLS) {
+      srxlTextBuff[row][col] = c;
+      lineSent[row] = false;
+    } else {
+      return 1;
+    }
+    return 0;
+}
+//**************************************************************************
+
+#define TEXT_GEN_KEEPALIVE_TIME_OUT 1000 // mS
+
+bool srxlFrameText(sbuf_t *dst, timeUs_t currentTimeUs)
+{
+  static timeUs_t lastTimeSent = 0;
+  static uint8_t lineNo = 0;
+  static uint8_t keepAliveLineNo = 0;
+
+  int lineCount = 0;
+  bool sent = false;
+
+  if ( (currentTimeUs - lastTimeSent)/1000 > TEXT_GEN_KEEPALIVE_TIME_OUT ) {
+    keepAliveLineNo = (keepAliveLineNo + 1) % SPEKTRUM_SRXL_DEVICE_TEXTGEN_ROWS;
+    lineNo = keepAliveLineNo;
+    lineSent[lineNo] = false;
+  }
+
+  // Skip already sent lines...
+  while (lineSent[lineNo] &&
+         lineCount < SPEKTRUM_SRXL_DEVICE_TEXTGEN_ROWS) {
+    lineNo = (lineNo + 1) % SPEKTRUM_SRXL_DEVICE_TEXTGEN_ROWS;
+    lineCount++;
+  }
+
+  if ( ! lineSent[lineNo] ) {
+    sbufWriteU8(dst, SPEKTRUM_SRXL_DEVICE_TEXTGEN);
+    sbufWriteU8(dst, SRXL_FRAMETYPE_SID);
+    sbufWriteU8(dst, lineNo);
+    for ( int i = 0; i < SPEKTRUM_SRXL_DEVICE_TEXTGEN_COLS; i++) {
+      sbufWriteU8(dst, srxlTextBuff[lineNo][i]);
+    }
+    lastTimeSent = currentTimeUs;
+    lineSent[lineNo] = true;
+    sent = true;
+    lineNo = (lineNo + 1) % SPEKTRUM_SRXL_DEVICE_TEXTGEN_ROWS;
+  }
+  return sent;
 }
 
+#endif
+
 // schedule array to decide how often each type of frame is sent
-#define SRXL_SCHEDULE_COUNT_MAX     3
+#define SRXL_SCHEDULE_MANDATORY_COUNT  2 // Mandatory QOS and RPM sensors
+
+#if defined (USE_SPEKTRUM_CMS_TELEMETRY) && defined (USE_CMS)
+#define SRXL_SCHEDULE_CMS_COUNT  1
+#else
+#define SRXL_SCHEDULE_CMS_COUNT  0
+#endif
+
+#define SRXL_SCHEDULE_USER_COUNT       (1 + SRXL_SCHEDULE_CMS_COUNT)
+
+#define SRXL_SCHEDULE_COUNT_MAX       3
```

Comment:
`SRXL_SCHEDULE_MANDATORY_COUNT  + 1` ? Otherwise it would skip user funcs/overflow
This scheduling is a bit confusing ...

Reply:
Confusing, yes. Added some more explanations and example.

Reply:
This frame scheduling and sequencing is not very well described in any Spektrum docs. I have some info on it in some old mail from Spektrum. It was also published on RCG: https://www.rcgroups.com/forums/showpost.php?p=36169367&postcount=1

To be absolutely sure there are no misunderstandings, or something I dreamt up, I did ask Spektrum if this is still valid, and I got a reply today confirming it.
 
    "Yes this is correct, not a dream :P"